src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalPopupMenu.java

```diff
@@ -127,11 +135,33 @@ public ToolbarButton getToolbarButton()
     */
    public void setActiveTerminal(String caption, String handle)
    {
-      activeTerminalCaption_ = caption;
       activeTerminalHandle_ = handle;
-      toolbarButton_.setText(trimCaption(activeTerminalCaption_));
+      String trimmed = trimCaption(caption);
+      if (handle != null)
+      {
+         trimmed = addBusyIndicator(trimmed, terminals_.getHasSubprocs(handle));
+      }
+      toolbarButton_.setText(trimmed);
    }
-   
+
+   /**
+    * Refresh caption of active terminal based on busy status.
```

Comment:
One way you can often avoid these kinds of "refresh UX to match state" functions is to use a `Value` wrapper around the state, and `addValueChangeHandler` to refresh the UX. This keeps you from getting into a state where the value has changed internally but isn't reflected in the UX. See: https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/studio/client/common/Value.java

No need to rewrite this, just something to keep in the back of your head!